crates/bevy_ecs/src/system/system_param.rs

```diff
@@ -525,45 +555,175 @@ impl<'w, 's> SystemParam for &'w World {
 }
 
 unsafe impl<'w, 's> SystemParamState for WorldState {
-    fn init(_world: &mut World, system_meta: &mut SystemMeta) -> Self {
-        let mut access = Access::default();
-        access.read_all();
+    fn init(world: &mut World, system_meta: &mut SystemMeta) -> Self {
+        // world can contain non-send resources, so it should probably be thread local
+        // system_meta.set_non_send();
+
+        let mut component_access = FilteredAccess::default();
+        let id = world.init_component::<WorldMetadata>();
+        component_access.add_read(id);
+        component_access.read_all();
+
         if !system_meta
+            .component_access_set
+            .get_conflicts(&component_access)
+            .is_empty()
+        {
+            panic!(
+                "&World conflicts with another system param in {}. \
+                Shared access and mutable access cannot overlap.",
+                system_meta.name,
+            );
+        }
+
+        system_meta.component_access_set.add(component_access);
+        // make sure executor sees conflict with another system having &mut World
+        system_meta
             .archetype_component_access
-            .is_compatible(&access)
+            .add_read(ArchetypeComponentId::WORLD_METADATA);
+        // prevent executor from concurrently running any systems that mutably borrow data
+        system_meta.archetype_component_access.read_all();
+
+        WorldState
+    }
+
+    // fn new_archetype(&mut self, archetype: &Archetype, system_meta: &mut SystemMeta) {
+    //     for info in archetype.components.values() {
+    //         system_meta.archetype_component_access.add_read(info.archetype_component_id);
+    //     }
+    // }
```

Comment:
Alternatively, we can avoid these confusing set theory questions by removing `reads_all` and `writes_all` and having the `&World` set bits like `Query` does.

Reply:
This gets my vote.

Reply:
So I actually think, as written, this would miss `Resource` types inserted late, since we'd never revisit the resource archetype. 

Maybe this was the real reason for `reads_all`. 

I don't think we can "fix" this without changing the signature of `SystemParamState` to have an `update_archetype_component_access` method.

Reply:
SGTM.